File: FlattenBTtoLL.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

public class FlattenBTtoLL {
//    class Solution {
//        Queue<TreeNode> q;
//        public void flatten(TreeNode root) {
//            if(root == null)
//                return;
//
//            this.q= new LinkedList<>();
//            perOrder(root);
//
//
//            TreeNode curr = root;
//            while(!q.isEmpty()){
//                TreeNode popped = q.poll();
//                if(popped == curr){
//                    curr.left = null;
//                }
//                else{
//                    curr.right = popped;
//                    curr.left = null;
//                    curr = curr.right;
//                }
//            }
//        }
//
//        private void perOrder(TreeNode root){
//            if(root == null)
//                return;
//
//            q.add(root);
//            perOrder(root.left);
//            perOrder(root.right);
//        }
//    }
class Solution {
    public void flatten(TreeNode root) {
        if(root == null || root.left == null && root.right ==null)
            return;

        if(root.left != null){
            flatten(root.left);
            TreeNode tempRight = root.right;
            root.right = root.left;
            root.left = null;
            while(root.right != null){
                root = root.right;
            }
            root.right = tempRight;
        }
        flatten(root.right);
    }
}
}

File: flatten-BT-to-LL.py
"""
Runtime Complexity:
O(n) - because we traverse through all the nodes.
Space Complexity:
O(n) - recrusive stack space. 
Yes, the worked on leetcode.
Issues while coding - No
"""

class Solution:
    def flattenTree(self, node):
        
        if (node==None):
            return 
        if (node.left==None) and (node.right==None):
            return node
        
        left = self.flattenTree(node.left)
        right =self.flattenTree(node.right)
        
        if left:
            left.right = node.right
            node.right = node.left
            node.left = None
        return right if right else left
    
    def flatten(self,root:TreeNode)-> None:
        self.flattenTree(root)